libappstream-builder/plugins/asb-plugin-appdata.c

```diff
@@ -57,16 +60,102 @@ asb_plugin_check_filename (AsbPlugin *plugin, const gchar *filename)
 	return _asb_plugin_check_filename (filename);
 }
 
+static GdkPixbuf *
+asb_app_load_icon (AsbPlugin *plugin,
+		   const gchar *filename,
+		   const gchar *logfn,
+		   guint icon_size,
+		   guint min_icon_size,
+		   GError **error)
+{
+	g_autoptr(AsImage) im = NULL;
+	g_autoptr(GError) error_local = NULL;
+	AsImageLoadFlags load_flags = AS_IMAGE_LOAD_FLAG_NONE;
+
+	/* is icon in a unsupported format */
+	if (!asb_context_get_flag (plugin->ctx, ASB_CONTEXT_FLAG_IGNORE_LEGACY_ICONS))
+		load_flags |= AS_IMAGE_LOAD_FLAG_ONLY_SUPPORTED;
+
+	im = as_image_new ();
+	if (!as_image_load_filename_full (im,
+					  filename,
+					  icon_size,
+					  min_icon_size,
+					  load_flags,
+					  &error_local)) {
+		g_set_error (error,
+			     ASB_PLUGIN_ERROR,
+			     ASB_PLUGIN_ERROR_FAILED,
+			     "%s: %s",
+			     error_local->message, logfn);
+		return NULL;
+	}
+	return g_object_ref (as_image_get_pixbuf (im));
+}
+
+static gboolean
+asb_plugin_convert_icons (AsbPlugin *plugin,
+			  const gchar *tmpdir,
+			  AsbApp *app,
+			  GError **error)
+{
+	AsIcon *icon;
+	GPtrArray *icons;
+	guint i;
+	guint min_icon_size;
+	g_autoptr(GdkPixbuf) pixbuf = NULL;
+	g_autofree gchar *name = NULL;
+
+	/* Convert local icons to cached.
+	 */
+
+	icons = as_app_get_icons (AS_APP (app));
+	for (i = 0; i < icons->len; i++) {
+		icon = g_ptr_array_index (icons, i);
+		if (as_icon_get_kind (icon) == AS_ICON_KIND_LOCAL) {
```

Comment:
What about the other icon kinds? e.g. `REMOTE` or `STOCK`? Shouldn't we add (without converting) those too?

Reply:
The code modifies the existing icons, so any non-local ones are left in place.

We could filter out (or reject) cached icons, though.